src/server/storage/mongo.js

```diff
@@ -330,47 +333,66 @@ function Mongo(mainLogger, gmeConfig) {
     }
 
     function openDatabase(callback) {
-        var deferred = Q.defer();
-        logger.debug('openDatabase');
-
-        if (mongo === null) {
-            logger.debug('Connecting to mongo...');
-            // connect to mongo
-            mongodb.MongoClient.connect(gmeConfig.mongo.uri, gmeConfig.mongo.options, function (err, db) {
-                if (!err && db) {
-                    mongo = db;
-                    logger.debug('Connected.');
-                    deferred.resolve();
-                } else {
-                    mongo = null;
-                    deferred.reject(err);
-                }
-            });
+        connectionCnt += 1;
+        logger.debug('openDatabase, connection counter:', connectionCnt);
+
+        if (connectionCnt === 1) {
+            if (mongo === null) {
+                logger.debug('Connecting to mongo...');
+                connectDeferred = Q.defer();
+                // connect to mongo
+                mongodb.MongoClient.connect(gmeConfig.mongo.uri, gmeConfig.mongo.options, function (err, db) {
+                    if (!err && db) {
+                        mongo = db;
+                        logger.debug('Connected.');
+                        connectDeferred.resolve();
+                    } else {
+                        mongo = null;
+                        connectionCnt -= 1;
+                        logger.error('Failed to connect.', {metadata: err});
+                        connectDeferred.reject(err);
+                    }
+                });
+            } else {
+                logger.debug('Count is 1 but mongo is not null');
+            }
         } else {
             logger.debug('Reusing mongo connection.');
```

Comment:
Who is resolving the connectDeferred here? Will the callback be called if the openDatabase is called twice?